src/Features/VisualBasic/Portable/CodeFixes/RemoveUnnecessaryCast/RemoveUnnecessaryCastCodeFixProvider.vb

```diff
@@ -70,117 +44,116 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeFixes.RemoveUnnecessaryCast
+            ' This is necessary to ensure that the code remains parseable and preserves semantics.
+            Dim expandedRoot = Await ExpandSurroundingStatementsAsync(trackedDocument, originalCastNodes, cancellationToken).ConfigureAwait(False)
+            Dim expandedDocument = document.WithSyntaxRoot(expandedRoot)
+
+            Dim removedRoot = Await RemoveCasts(
+                expandedDocument, originalCastNodes, cancellationToken).ConfigureAwait(False)
+
+            editor.ReplaceNode(editor.OriginalRoot, removedRoot)
+        End Function
 
-            updatedDocument = Await updatedDocument.ReplaceNodeAsync(parentStatement, explicitParentStatement, cancellationToken).ConfigureAwait(False)
+        Private Async Function RemoveCasts(
+                document As Document, originalCastNodes As ImmutableArray(Of ExpressionSyntax),
+                cancellationToken As CancellationToken) As Task(Of SyntaxNode)
 
-            ' Next, make the statement after the parenting statement of the expression semantically explicit.
-            parentStatement = Await FindNodeWithAnnotationAsync(Of StatementSyntax)(s_statementAnnotation, updatedDocument, cancellationToken).ConfigureAwait(False)
-            Dim nextStatement = parentStatement.GetNextStatement()
-            If nextStatement IsNot Nothing Then
-                Dim explicitNextStatement = Await Simplifier.ExpandAsync(nextStatement, updatedDocument, cancellationToken:=cancellationToken).ConfigureAwait(False)
-                updatedDocument = Await updatedDocument.ReplaceNodeAsync(nextStatement, explicitNextStatement, cancellationToken).ConfigureAwait(False)
-            End If
+            Dim root = Await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(False)
 
-            updatedDocument = Await RewriteCoreAsync(updatedDocument, expression, cancellationToken).ConfigureAwait(False)
+            ' Now, find the cast nodes again in the expanded document
+            Dim currentCastNodes = root.GetCurrentNodes(originalCastNodes)
 
-            ' Remove added _expressionAnnotation and _statementAnnotation.
-            updatedDocument = Await RemoveNodesAndTokensWithAnnotationAsync(s_expressionAnnotation, updatedDocument, cancellationToken).ConfigureAwait(False)
-            updatedDocument = Await RemoveNodesAndTokensWithAnnotationAsync(s_statementAnnotation, updatedDocument, cancellationToken).ConfigureAwait(False)
+            Dim innerEditor = New SyntaxEditor(root, document.Project.Solution.Workspace)
+            Await innerEditor.ApplyExpressionLevelSemanticEditsAsync(
+                document, currentCastNodes.ToImmutableArray(),
+                Function(semanticModel, castExpression) IsUnnecessaryCast(castExpression, semanticModel, cancellationToken),
+                Function(unused, currentRoot, castExpression)
+                    Dim newCastExpression = Uncast(castExpression).WithAdditionalAnnotations(Formatter.Annotation)
```

Comment:
so since it already did complexify, it doesn't do what C# does I guess. can C# do what VB does and have same implementation? not doing worry unnecessary complexify increase perf cost?

Reply:
it could... but i'm wary about adding unnecessary steps for C#.  VB needs this because VB is hugely ambiguous at the statement level.  i.e. it's not at all clear where a VB statement begins/ends, and adding/removing parens can change the parser's view on htat.  For example, VB might end a statement at the end of a line if you had ```DirectCast(...)``` as the last thing.  However, when you get rid of that, it might then add an implicit-line-continuation and absorb the next line into the previous one.

Similarly, casts are fine to start statements in VB. But the expression being casted is not necessarily fine to start a statement.  Complexify solves both these problems.

C# has neither of these problems.  If a casted expression can start an expr-statement, then the expr being casted can start a expr-statement.  Similarly, C# statements always have a well defined end (due to ```;``` or ```}```) so there is no concern about the next statement being 'pulled into' the previous one.

I'm personally ok with VB having this extra complexity because we're mutating the parse tree and hte VB parse tree is just generally much more complex and ambiguous when you touch the start/end of any statement.  For C# that just isn't a problem, so i would prefer to not have to add extra code there.

Reply:
I am fine having it only in VB.